pkg/agent/controller/multicast/mcast_controller.go

```diff
@@ -0,0 +1,200 @@
+			m.mutex.RLock()
+			obj, ok, _ := m.groupCache.GetByKey(e.group.String())
+			m.mutex.RUnlock()
+			switch e.eType {
+			case groupJoin:
+				if !ok {
+					m.addGroupMemberStatus(e)
+				} else {
+					m.updateGroupMemberStatus(obj, e)
+				}
+			case groupLeave:
+				if ok {
+					m.updateGroupMemberStatus(obj, e)
+				}
+			}
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// addGroupMemberStatus adds the new group into groupCache.
+func (m *MulticastController) addGroupMemberStatus(e mcastGroupEvent) error {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	status := &GroupMemberStatus{
+		group:        e.group,
+		lastProbe:    e.time,
+		localMembers: map[string]bool{e.iface.InterfaceName: true},
+	}
+	m.groupCache.Add(status)
+	klog.InfoS("Add new Multicast group to cache", "group", e.group, "interface", e.iface.InterfaceName)
+	return nil
+}
+
+// updateGroupMemberStatus updates the group status in groupCache. If a "join" message is sent from an existing member,
+// only updates the lastProbe time. If a "join" message is sent from an "unknown" member, updates the lastProbe time and
+// adds it into the group's local member set. If a "leave" message is sent from an existing member, remove it from the
+// group's local member set, and if the member is the last one in local cache, a query message on the group is sent out
+// to check if it is the last one on OVS.
+func (m *MulticastController) updateGroupMemberStatus(obj interface{}, e mcastGroupEvent) error {
+	status := obj.(*GroupMemberStatus)
+	m.mutex.RLock()
+	_, exist := status.localMembers[e.iface.InterfaceName]
+	m.mutex.RUnlock()
+	switch e.eType {
+	case groupJoin:
+		status.mutex.Lock()
+		status.lastProbe = e.time
+		if !exist {
+			// TODO: add (Pod_IP, e.group) in the Multicast Routing entry whose inbound interface is antrea-gw0
+			status.localMembers[e.iface.InterfaceName] = true
+			klog.InfoS("Add interface to Multicast group", "group", e.group.String(), "member", e.iface.InterfaceName)
+		}
+		status.mutex.Unlock()
+		m.mutex.Lock()
+		m.groupCache.Update(status)
+		m.mutex.Unlock()
+	case groupLeave:
+		if exist {
+			// TODO: remove (Pod_IP, e.group) in the Multicast Routing entry whose inbound interface is antrea-gw0
+			status.mutex.Lock()
+			delete(status.localMembers, e.iface.InterfaceName)
+			status.mutex.Unlock()
+			klog.InfoS("Member left the Multicast group", "group", e.group.String(), "member", e.iface.InterfaceName)
+			if len(status.localMembers) == 0 {
+				klog.InfoS("Check last member in Multicast group to cache", "group", e.group.String(), "member", e.iface.InterfaceName)
+				go m.checkLastMember(status, e.group)
+			}
+		}
+	}
+	return nil
+}
+
+// checkLastMember sends out a query message on the group to check if the last member on the group left. If no new
+// membership report is received in the max response time, the group is removed from groupCache.
+func (m *MulticastController) checkLastMember(status *GroupMemberStatus, group net.IP) {
```

Comment:
Maybe we can use `context.WithTimeout` here.
Like:
```go
ctx, cancelFunc := context.WithTimeout(context.Background(), time.Second)
defer cancelFunc()
m.groupDetector.queryIGMP(cancelFunc, group, queryVersion)

<-ctx.Done
if err := ctx.Error(); err != nil {
    // handle error.
}
// positive path.
```

Reply:
Thanks for the suggestion. Think over, I still feel it is not necessary to introduce a single context for each checkLastMember, for the context cost is larger than using time.After directly. But I use a context in the MulticastController, and if the MulticastController is stopped, all the goroutines it has created for "checkLastMember" can be notified to return. What do you think?

As for the unit test, I change the response time in test code to cut off the execution time.